doc/dev/osd_internals/manifest.rst

```diff
@@ -110,6 +187,49 @@ cache/tiering promotion to ensure that the ObjectStore ends up with
 the right byte range sharing via clone.  We probably want to ensure
 that adjacent clones agree on chunk mappings for shared regions.
 
+* Implementation
+
+  My thought is that set_chunk and set_reredirect shouldn't set
+  user_modify to indicate not to trigger a new clone (probably also need a flag like cache_evict).
+  Also, to use manifest with snapshot, set_chunk should be set when the manifest object is created,
+  and disallow to make a new clone on the manifest object which already has clones.
+  As a result, the overall procedure is:
+  1. Write the object A
+  2. SET_CHUNK (offset: 0 ~ 4)
+  3. SET_CHUNK (offset: 8 ~ 12)
+  4. SET_CHUNK (offset: 16 ~ 20)
+  5. Create a snapshot 
+  6. Write the object A
+  7. Create a snapshot
+  When a snapshot is created between 5 and 6, clones prior to head are dirty
+  if the head object is dirty (and the flush should start from old clones).
+
+  The state of the chunk_map in clone should be MISSING after eviction is done.
+  If we want to read clones, we have to look at the chunk_info_t according to given snap_id 
+  to find out which chunk is needed.
+  Therefore, clone needs to be mutable. To do that, we probably need a new op type.
+  If snap read occurs, the manifest object might have the overwritten chunks because
+  the contents are different depending on what snap_id is.
```

Comment:
I'm not sure I understand this.  A snap will always map to a particular clone, which would have its own manifest.  I think you just need to clean up the osd code to ensure that SET_CHUNK and friends operate validly on a clone, and where appropriate maintain clone_overlap.  I'd like a bit more detail here on how that would work, in particular, on how you'd maintain clone_overlap (note that failing to account for this will result in the osd crashing or corrupting data during recovery).

Reply:
please see the comment above.

Reply:
I think you still haven't really addressed the clone_overlap component.  If you are referring to "I mean "disallow to make a new chunk which already has clones via SETCHUNK"", then I think that's problematic.  Consider that clones of an active rbd image would be the perfect dedup targets.  In that case, SET_CHUNK would never actually get called on the HEAD, only on the clones.  In that case,  you really would have to deal with clone_overlap for adjacent clones.

Reply:
Ok. see below comment.